contracts/0.8.9/WithdrawalRequestNFT.sol

```diff
@@ -0,0 +1,300 @@
+        request.owner = payable(_to);
+
+        _getRequestsByOwner()[_to].add(_requestId);
+        _getRequestsByOwner()[_from].remove(_requestId);
+
+        _emitTransfer(_from, _to, _requestId);
+    }
+
+    /// @dev Internal function to invoke {IERC721Receiver-onERC721Received} on a target address.
+    /// The call is not executed if the target address is not a contract.
+    ///
+    /// @param _from address representing the previous owner of the given token ID
+    /// @param _to target address that will receive the tokens
+    /// @param _requestId uint256 ID of the token to be transferred
+    /// @param _data bytes optional data to send along with the call
+    /// @return bool whether the call correctly returned the expected magic value
+    function _checkOnERC721Received(address _from, address _to, uint256 _requestId, bytes memory _data)
+        private
+        returns (bool)
+    {
+        if (_to.isContract()) {
+            try IERC721Receiver(_to).onERC721Received(msg.sender, _from, _requestId, _data) returns (bytes4 retval) {
+                return retval == IERC721Receiver.onERC721Received.selector;
+            } catch (bytes memory reason) {
+                if (reason.length == 0) {
+                    revert TransferToNonIERC721Receiver(_to);
+                } else {
+                    /// @solidity memory-safe-assembly
+                    assembly {
+                        revert(add(32, reason), mload(reason))
+                    }
+                }
+            }
+        } else {
+            return true;
+        }
+    }
+
+    //
+    // Internal getters and setters
+    //
+
+    /// @dev a little crutch to emit { Transfer } on request and on claim like ERC721 states
+    function _emitTransfer(address _from, address _to, uint256 _requestId) internal override {
+        emit Transfer(_from, _to, _requestId);
+    }
+
+    /// @dev Returns whether `_requestId` exists and not claimed.
+    function _existsAndNotClaimed(uint256 _requestId) internal view returns (bool) {
+        return _requestId > 0 && _requestId <= getLastRequestId() && !_getQueue()[_requestId].claimed;
+    }
+
+    /// @dev Approve `_to` to operate on `_requestId`
+    /// Emits a { Approval } event.
+    function _approve(address _to, uint256 _requestId) internal {
+        _getTokenApprovals()[_requestId] = _to;
+        emit Approval(ownerOf(_requestId), _to, _requestId);
+    }
+
+    /// @dev Approve `operator` to operate on all of `owner` tokens
+    /// Emits a { ApprovalForAll } event.
+    function _setApprovalForAll(address _owner, address _operator, bool _approved) internal {
+        if (_owner == _operator) revert ApproveToCaller();
+        _getOperatorApprovals()[_owner][_operator] = _approved;
+        emit ApprovalForAll(_owner, _operator, _approved);
+    }
+
+    /// @dev Decode a `bytes32 to string
+    function _toString(bytes32 _sstr) internal pure returns (string memory) {
```

Comment:
These function `_toString, _toBytes32, _length` look aliens taken from somewhere. Maybe add comments about the source? Move them to a base util contact? Just refactor naming to match the contarct?